pkg/controllers/provisioning/scheduling/scheduler.go

```diff
@@ -75,52 +72,30 @@ func (s *Scheduler) Solve(ctx context.Context, provisioner *v1alpha5.Provisioner
 		return nil, fmt.Errorf("injecting topology, %w", err)
 	}
 	// Separate pods into schedules of isomorphic scheduling constraints.
-	schedules, err := s.getSchedules(ctx, constraints, pods)
-	if err != nil {
-		return nil, fmt.Errorf("getting schedules, %w", err)
-	}
+	schedules := s.getSchedules(constraints, pods)
 	return schedules, nil
 }
 
 // getSchedules separates pods into a set of schedules. All pods in each group
 // contain isomorphic scheduling constraints and can be deployed together on the
 // same node, or multiple similar nodes if the pods exceed one node's capacity.
-func (s *Scheduler) getSchedules(ctx context.Context, constraints *v1alpha5.Constraints, pods []*v1.Pod) ([]*Schedule, error) {
-	// schedule uniqueness is tracked by hash(Constraints)
-	schedules := map[uint64]*Schedule{}
+func (s *Scheduler) getSchedules(constraints *v1alpha5.Constraints, pods []*v1.Pod) []*Schedule {
+	schedules := []*Schedule{}
 	for _, pod := range pods {
-		if err := constraints.ValidatePod(pod); err != nil {
-			logging.FromContext(ctx).Infof("Unable to schedule pod %s/%s, %s", pod.Namespace, pod.Name, err)
-			continue
-		}
-		tightened := constraints.Tighten(pod)
-
-		// schedulingConstraints applies the provisioner constraints
-		// and any inferred constraints such as GPU resource requests from the pods
-		// and is then hashed to compute the schedules
-		schedulingConstraints := struct {
-			*v1alpha5.Constraints
-			GPURequests v1.ResourceList
-		}{
-			Constraints: tightened,
-			GPURequests: resources.GPULimitsFor(pod),
+		isCompatible := false
+		for index, schedule := range schedules {
+			if err := schedule.Requirements.Compatible(v1alpha5.NewPodRequirements(pod)); err == nil {
+				//TODO: The cross product of the constraints values from cloud provider may not be supported by any instance type.
```

Comment:
Can you cut an issue for this?

Reply:
This multi-dimentional matching problem is actually the core of the scheduling problem that make it NP-complete. We are avoiding it because of some strong assumptions and relaxations that allows us to decouple pod grouping and node selection. The virtual node solution will allow us to properly model this process.

Optimally, the solution workflow will be like:
1. Given the current instance types selection for the virtual nodes, find pod groupings.
2. Adjust the  instance type selections (e.g., number of nodes, switch to a compatible type) to optimize the objective
3. Recalculate step 1 to see the new  instance type selections in step 2 is feasible.
Repeat until converge.

The above validation in line 87 is actually a simplified version of step 2 + 3.
The above iterative approach can be extremely expensive. Luckily cloud providers like AWS has quite a collection of homogenous instance type with different size in each instance type groups. We should make use of that property and do some heuristic design to short cut step 3. 

The virtual node approach will be the core of the future consolidation solution as well. Let's spend some time to go over the design.